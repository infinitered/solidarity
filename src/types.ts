import { GluegunRunContext } from 'gluegun'
export const solidarity = {
  binaryExists: require('./extensions/functions/binaryExists'),
  getSolidaritySettings: require('./extensions/functions/getSolidaritySettings'),
  setSolidaritySettings: require('./extensions/functions/setSolidaritySettings'),
  checkRequirement: require('./extensions/functions/checkRequirement'),
  updateRequirement: require('./extensions/functions/updateRequirement'),
  skipRule: require('./extensions/functions/skipRule'),
  checkENV: require('./extensions/functions/checkENV'),
  checkCLI: require('./extensions/functions/checkCLI'),
  checkDir: require('./extensions/functions/checkDir'),
  checkFile: require('./extensions/functions/checkFile'),
  updateVersions: require('./extensions/functions/updateVersions'),
  removeNonVersionCharacters: require('./extensions/functions/removeNonVersionCharacters'),
  getVersion: require('./extensions/functions/getVersion'),
  getLineWithVersion: require('./extensions/functions/getLineWithVersion')
}

export interface SolidarityPlugin {
  name: string
  description: string
  snapshot: string | SnapshotType
}

export interface SolidarityRunContext extends GluegunRunContext {
  solidarity: typeof solidarity
  _pluginsList: Array<SolidarityPlugin & {templateDirectory: string}>
  addPlugin: (config: SolidarityPlugin) => void
  printSeparator: () => void
  outputMode: SolidarityOutputMode
}

export type SnapshotType = (context: SolidarityRunContext) => Promise<void>

export interface SolidarityRule {
  rule: 'cli' | 'env' | 'dir' | 'file'
  binary?: string
  semver?: string
  version?: string
  line?: string | number
  variable?: string
  location?: string
  error?: string
  platform?: string | string[]
  matchIndex?: number
}

export type SolidarityRequirement = SolidarityRule[]

export interface SolidaritySettings {
  requirements: object
  config: object
}

export enum SolidarityOutputMode {
  MODERATE,
  VERBOSE,
  SILENT
}
