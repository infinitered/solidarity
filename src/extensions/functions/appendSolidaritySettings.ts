import { keys } from 'ramda'

module.exports = (solidaritySettings, newRequirement) => {
  const newRequirementKey = keys(newRequirement)
  const existingRequirementRules = solidaritySettings.requirements[newRequirementKey] || []

  return {
    ...solidaritySettings,
    requirements: {
      ...solidaritySettings.requirements,
      [newRequirementKey]: [
        ...existingRequirementRules,
        ...newRequirement[newRequirementKey]
      ]
    }
  }
}
