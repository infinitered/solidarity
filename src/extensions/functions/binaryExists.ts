import { SolidarityRunContext } from '../../types'
module.exports = (binary: string, context: SolidarityRunContext): boolean => {
  const { system } = context

  // Check if binary exists
  try {
    system.which(binary)
    return true
  } catch (_e) {
    return false
  }
}
