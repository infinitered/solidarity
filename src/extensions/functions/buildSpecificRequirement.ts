import { SolidarityRunContext, SolidaritySettings } from '../../types'
import { keys } from 'ramda'

namespace buildSpecificRequirement {
  const requiredInputQuestion = async ({ name, message, prompt }) => {
    const result = await prompt.ask({
      name,
      type: 'input',
      message,
    })

    if (!result[name]) return Promise.reject('An input is required')
    return result
  }

  const resolveParameters = async ({ parameters, prompt, ruleHandlers }) => {
    const { first, second } = parameters
    if (second) {
      return Promise.resolve({ whatRule: second })
    }

    const message = `What's the ${first} ${ruleHandlers[first].key} you'd like to add a rule for?`

    return requiredInputQuestion({
      name: 'whatRule',
      message,
      prompt,
    })
  }

  const getRequirementNames = (solidaritySettings: SolidaritySettings): Array<string> =>
    keys(solidaritySettings.requirements)

  const chooseRequirement = async (prompt, solidaritySettings: SolidaritySettings): Promise<String> => {
    const shouldMakeNewRequirement = await prompt.ask({
      name: 'makeNewRequirement',
      type: 'confirm',
      message: 'Would you like to create a new requirement set?',
    })

    let requirementName

    if (shouldMakeNewRequirement.makeNewRequirement) {
      const answer = await requiredInputQuestion({
        name: 'newRequirement',
        message: 'What would you like to call this new requirement?',
        prompt,
      }).catch(error => {
        return Promise.reject(error)
      })
      requirementName = answer.newRequirement
    } else {
      const requirementOptions = getRequirementNames(solidaritySettings)
      const answer = await prompt.ask({
        name: 'selectedRequirement',
        message: 'Which of the above technology snapshots will you use for this project?',
        type: 'selet',
        choices: requirementOptions,
      })
      requirementName = answer.selectedRequirement
    }

    return requirementName
  }

  const constructRequirment = async context => {
    const { parameters, prompt, solidarity } = context
    const { getSolidaritySettings, ruleHandlers } = solidarity
    const solidaritySettings = await getSolidaritySettings(context)

    const userAnswer = await prompt.ask({
      name: 'addNewRule',
      type: 'confirm',
      message: `Would you like to add the ${parameters.first} '${parameters.second}' to your Solidarity file?`,
    })

    if (userAnswer.addNewRule) {
      // maybe ask about setting up the new rule w/ a specific version?
      const requirementName = await chooseRequirement(prompt, solidaritySettings)
      return ruleHandlers[parameters.first].callback(context, requirementName)
    } else {
      return Promise.reject('Rule not added.')
    }
  }

  export const run = async (context: SolidarityRunContext) => {
    const { parameters, prompt, solidarity } = context
    const { first } = parameters
    const { ruleHandlers } = solidarity
    const resolvedParam = await resolveParameters({ parameters, prompt, ruleHandlers }).catch(() => {
      return Promise.reject('Missing required parameters.')
    })

    return constructRequirment({
      ...context,
      parameters: {
        ...parameters,
        first,
        second: resolvedParam.whatRule,
      },
    })
  }
}

module.exports = buildSpecificRequirement.run
