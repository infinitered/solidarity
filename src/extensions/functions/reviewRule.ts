import {
  SolidarityRequirementChunk,
  SolidarityRequirement,
  SolidarityRunContext,
  SolidarityRule,
  SolidarityReportResults,
} from '../../types'
import { tail, pipe, flatten, map } from 'ramda'
const skipRule = require('./skipRule')
const checkDir = require('./checkDir')
const checkFile = require('./checkFile')
const checkShell = require('./checkShell')
const findPluginInfo = require('./findPluginInfo')

module.exports = async (
  requirement: SolidarityRequirementChunk,
  report: SolidarityReportResults,
  context: SolidarityRunContext
) => {
  const { print, solidarity } = context
  const { colors, checkmark, xmark } = print
  const prettyBool = (bl: boolean) => (bl ? checkmark + colors.green(' YES') : xmark + colors.red(' NO'))
  // @ts-ignore - flatten will never get a string bc tail is called first
  const rules: SolidarityRequirement = pipe(tail, flatten)(requirement)
  // check each rule for report
  const ruleChecks = map(async (rule: SolidarityRule) => {
    // Make sure this rule is active
    if (skipRule(rule.platform)) return false

    switch (rule.rule) {
      // Handle CLI rule report
      case 'cli':
        let binaryVersion
        try {
          binaryVersion = await solidarity.getVersion(rule, context)
        } catch (_e) {
          binaryVersion = colors.red('*UNKNOWN*')
        }

        report.addCLI({
          binary: rule.binary,
          version: binaryVersion,
          desired: rule.semver,
        })
        break
      // Handle ENV rule report
      case 'env':
        const envValue = process.env[rule.variable] || colors.red('*UNDEFINED*')
        report.envRules.push([`$${rule.variable}`, envValue])
        break
      // Handle dir rule report
      case 'directory':
      case 'dir':
        const dirExists = prettyBool(checkDir(rule, context))
        report.filesystemRules.push([rule.location, 'Dir', dirExists])
        break
      // Handle file rule report
      case 'file':
        const fileExists = prettyBool(checkFile(rule, context))
        report.filesystemRules.push([rule.location, 'File', fileExists])
        break
      case 'shell':
        const shellCheckPass = prettyBool(await checkShell(rule, context))
        report.shellRules.push([rule.command, rule.match, shellCheckPass])
        break
      case 'custom':
        const customPluginRule = findPluginInfo(rule, context)
        if (customPluginRule.success) {
          // let plugin update the report
          if (customPluginRule.plugin.report) customPluginRule.plugin.report(rule, context, report)
        } else {
          throw new Error(customPluginRule.message)
        }
        break
      default:
        throw new Error('Encountered unknown rule')
    }
  }, rules)

  // Run all the rule checks for a requirement
  return Promise.all(ruleChecks).then(results => {
    return results
  })
}
