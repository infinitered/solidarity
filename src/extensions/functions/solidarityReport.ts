import { SolidarityRunContext, SolidarityReportResults, CLIReportConfig } from '../../types'

export const createReport = (context: SolidarityRunContext): SolidarityReportResults => {
  const { print, system, envinfoHelpers } = context
  const { colors } = print

  return {
    basicInfo: [
      ['System Basics', 'Value'],
      ['OS', envinfoHelpers.getOperatingSystemInfo()],
      ['CPU', envinfoHelpers.getCPUInfo()],
    ],
    cliRules: [['Binary', 'Location', 'Version', 'Desired']],
    envRules: [['Environment Var', 'Value']],
    filesystemRules: [['Location', 'Type', 'Exists']],
    shellRules: [['Command', 'Pattern', 'Matches']],
    customRules: [],
    // helper for adding CLI rules
    addCLI: function(cliReportConfig: CLIReportConfig) {
      const desired = cliReportConfig.desired ? cliReportConfig.desired : colors.green('*ANY*')
      let location
      try {
        location = system.which(cliReportConfig.binary)
      } catch (_e) {
        location = colors.red('*MISSING*')
      }
      this.cliRules.push([cliReportConfig.binary, location, cliReportConfig.version, desired])
    },
  }
}
