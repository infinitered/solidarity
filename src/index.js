const { build } = require('gluegun')
const {dependenciesMet} = require('gluegun/sniff')
if(dependenciesMet==1){
    module.exports = async () => {
      // setup the runtime
      build()
        .brand('solidarity')
        .src(__dirname)
        // It might also be worth checking global
        // `npm config get prefix` returns '/usr/local'
        // global npm is stored /usr/local/lib/node_modules
        .plugins('./node_modules', { matching: 'solidarity-*', hidden: true })
        // for testing - force load a plugin
        // .plugin('../solidarity-react-native')
        .create()
        .run()
    }
}
