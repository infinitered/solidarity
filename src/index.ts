import { build } from 'gluegun'

module.exports = async (): Promise<void> => {
  // setup the runtime
  build()
    .brand('solidarity')
    .src(__dirname)
    // local installs
    .plugins('./node_modules', { matching: 'solidarity-*', hidden: true })
    // global installs
    .plugins('/usr/local/lib/node_modules', { matching: 'solidarity-*', hidden: true })
    // for testing - force load a local plugin
    // .plugin('../solidarity-react-native')
    .create()
    .run()
}

export * from './types'
