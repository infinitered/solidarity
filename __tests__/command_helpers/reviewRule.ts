import reviewRule from '../../src/extensions/functions/reviewRule'
import { SolidarityRunContext, SolidarityReportResults } from '../../src/types'
import { createReport } from '../../src/extensions/functions/solidarityReport'
const examplePlugin = require('examplePlugin')
let mockContext: SolidarityRunContext
let reportResults: SolidarityReportResults
describe('reviewRule', () => {
  beforeEach(() => {
    // fresh mock context
    mockContext = examplePlugin(require('mockContext'))
    reportResults = createReport(mockContext)
  })

  describe('when rule: cli', () => {
    test('rule gets added', async () => {
      const rule = ['NPM', [{ rule: 'cli', binary: 'npm' }]]

      const result = await reviewRule(rule, reportResults, mockContext)
      // CLI rule was added
      expect(reportResults.cliRules.length).toBe(2)
    })
  })

  describe('when rule: env', () => {
    test('rule gets added', async () => {
      const rule = ['ANDROID', [{ rule: 'env', value: 'ANDROID_HOME' }]]

      const result = await reviewRule(rule, reportResults, mockContext)
      // CLI rule was added
      expect(reportResults.envRules.length).toBe(2)
    })
  })

  describe('when rule: dir', () => {
    test('rule gets added', async () => {
      const rule = ['DIRECTORY', [{ rule: 'dir', binary: 'random' }]]

      const result = await reviewRule(rule, reportResults, mockContext)
      // CLI rule was added
      expect(reportResults.filesystemRules.length).toBe(2)
    })
  })

  describe('when rule: file', () => {
    test('rule gets added', async () => {
      const rule = ['FILE', [{ rule: 'file', binary: 'random' }]]

      const result = await reviewRule(rule, reportResults, mockContext)
      // CLI rule was added
      expect(reportResults.filesystemRules.length).toBe(2)
    })
  })

  describe('when rule: shell', () => {
    test('rule gets added', async () => {
      const rule = ['SHELL', [{ rule: 'shell', command: 'ls', match: '.+' }]]

      const result = await reviewRule(rule, reportResults, mockContext)
      // SHELL rule was added
      expect(reportResults.shellRules.length).toBe(2)
    })
  })

  // Custom rule test
  describe('when rule: custom', () => {
    test('rule gets added', async () => {
      const rule = ['CUSTOM', [{ rule: 'custom', plugin: 'Example Plugin', name: 'checkThing' }]]

      expect(reportResults.cliRules.length).toBe(1)
      const result = await reviewRule(rule, reportResults, mockContext)
      // CUSTOM rule (which adds CLI report) was added
      expect(reportResults.cliRules.length).toBe(2)
    })

    test('does nothing when no report exists', async () => {
      const rule = ['CUSTOM', [{ rule: 'custom', plugin: 'Example Plugin', name: 'checkSecondThing' }]]

      expect(reportResults.cliRules.length).toBe(1)
      const result = await reviewRule(rule, reportResults, mockContext)
      // should not change rules
      expect(reportResults.cliRules.length).toBe(1)
    })

    test('Errors when plugin doesn not exist', async () => {
      const rule = ['CUSTOM', [{ rule: 'custom', plugin: 'FAKE', name: 'checkSecondThing' }]]

      // Async error snapshots (not simple)
      try {
        await reviewRule(rule, reportResults, mockContext)
        fail('Unknown rule should have errored')
      } catch (e) {
        expect(e).toMatchSnapshot()
      }
    })
  })

  describe('when rule: unknown', () => {
    test('rule gets added', async () => {
      const rule = ['UNKNOWN', [{ rule: 'UNKNOWN', command: 'ls', match: '.+' }]]

      // Async error snapshots (not simple)
      try {
        await reviewRule(rule, reportResults, mockContext)
        fail('Unknown rule should have errored')
      } catch (e) {
        expect(e).toMatchSnapshot()
      }
    })
  })
})
