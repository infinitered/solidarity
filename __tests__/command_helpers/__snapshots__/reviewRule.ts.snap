// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reviewRule when rule: custom Errors when plugin doesn not exist 1`] = `[Error: Plugin not found 'FAKE']`;

exports[`reviewRule when rule: unknown rule gets added 1`] = `[Error: Encountered unknown rule]`;
